### DISCUSSION TITLE
Finance::Quote Documentation in GnuCash

### DISCUSSION THREAD

**Discussion by bpschuck** [finance-quote/finance-quote#423] [General]
The documentation about Finance::Quote in GnuCash has been out of date for quite some time. In addition, the list of F::Q sources in the GnuCash program has also lagged behind changes to F::Q.

In the GnuCash manual, the out of date section is "A.1. Finance::Quote Sources", the direct link to the HTML is https://www.gnucash.org/docs/v5/C/gnucash-manual/fq-sources.html. In the GnuCash Docs repository, the file that needs to be updated to match F::Q is `C/manual/tips-appendix.docbook`.

In the GnuCash itself, the file that needs to be kept in-sync with F::Q is `libgnucash/engine/gnc-commodity.cpp`.

While these files should be easy to maintain once brought up to date, the work to bring them up to date with current F::Q modules and methods is more involved. It would be appreciated if someone volunteers to tackle the initial updates.

**Comment by pkryger** [comment]
I am not sure if this is exactly how it has been envisioned, but I created the following  PRs for this work: https://github.com/Gnucash/gnucash/pull/2033 and https://github.com/Gnucash/gnucash-docs/pull/345. Please let me know if you have any comments.

**Comment by bpschuck** [comment]
Thank you for these. I appreciate the contribution. When I have time, it's been spent on addressing issues and pull requests.
I haven't added it to the Hacker's Guide yet, but I've been asking people fixing or updating modules to incorporate the `methodinfo` method. This changes how `labels` and `methods` are defined. See `FinanceAPI.pm`

```
our $DISPLAY    = 'FinanceAPI';
our $FEATURES   = {'API_KEY' => 'registered user API key'};
our @LABELS     = qw/symbol name open high low last price date volume currency method/;
our $METHODHASH = {subroutine => \&financeapi, 
                   display => $DISPLAY, 
                   labels => \@LABELS,
                   features => $FEATURES};

sub methodinfo {
    return ( 
        financeapi   => $METHODHASH,
        nyse         => $METHODHASH,
        nasdaq       => $METHODHASH,
        usa          => $METHODHASH,
    );
}

sub labels { my %m = methodinfo(); return map {$_ => [@{$m{$_}{labels}}] } keys %m; }

sub methods {
  my %m = methodinfo(); return map {$_ => $m{$_}{subroutine} } keys %m;
}
```

Having just merged a PR that fixed Comdirect.pm, only to have them change the format of the HTML/CSS returned not even 24 hours afterwards, I've started to think of adding a field to Modules-README.yml that indicates the stability of the module, how often it's needed to be changed to account for changes. Maybe even present that in the `methodinfo` hash as well.

Thank you.

Bruce S.

**Reply by pkryger** [reply]
> Thank you for these. I appreciate the contribution.

You're welcome ☺️.

> [...] I haven't added it to the Hacker's Guide yet, but I've been asking people fixing or updating modules to incorporate the `methodinfo` method. [...]

I gather this is an ask for #439. I updated it to have the `methodinfo`.

> Having just merged a PR that fixed Comdirect.pm, only to have them change the format of the HTML/CSS returned not even 24 hours afterwards, I've started to think of adding a field to Modules-README.yml that indicates the stability of the module, how often it's needed to be changed to account for changes. 

I can totally understand you. It's really frustrating to learn that the code you've been working on has been rendered invalid by some other change 😤.

> Maybe even present that in the `methodinfo` hash as well.

It seems like a good idea. Perhaps more of the data that is currently  captured in `Modules-README.yml` (all of it?) could be present in `methodinfo` 🤔. That way most of test cases that fetch data from outside sources could be written automatically. Perhaps only keep Modules-README.yml as a graveyard for modules that have been removed. This may be useful for a case when someone adds a new module with a name that had been used in the past, but the module is for something else than the dead one. Given how Gnucash I guess it may cause confusion: to my understanding `Finance::Quote`'ss source names are stored in user files (`.gnucash`) and whatever version of Gnucash that is trying to pull a fresh quotes is just blindly sending them to `Finance::Quote`. If information of past modules is present such a situation can be caught in an automatic way.

Cheers,
PK

**Comment by bpschuck** [comment]
@pkryger,

`methodinfo` is only meant as a tool to enable future versions of GnuCash dynamically create the commodities now hardcoded into the `commodities.cpp` file. Not as a tool to help create and update documentation.And since it would need to stay for keeping track of removed modules, why not continue using it as is.

WRT to the GnuCash manual appendix doc, do you understand what "multiple sources" are? They are not countries, regions, etc. F::Q has a built-in "failover" mechanism. If one were to have "nyse" as the source in the GC security editor, F::Q would try the "nyse" method in each module that advertises it until one successfully returns data. 

Technically, while the `Stooq.pm` module advertises `stooq` and `poland` as methods, since no other module supplies a `poland` method, then `poland` is not a multiple source. It could be sometime in the future, but not currently. 

The only methods listed in that table should be those **currently** available in multiple modules.

Thank you.

**Comment by pkryger** [comment]
I ported the tables to GnuCash Wiki, please see https://wiki.gnucash.org/wiki/Online_Quotes#Finance::Quote_Sources and the following section.

**Comment by bpschuck** [comment]
On 10/22/24 05:46, Przemysław Kryger wrote:

> I ported the tables to GnuCash Wiki, please see 
> https://wiki.gnucash.org/wiki/Online_Quotes#Finance::Quote_Sources 
> <https://wiki.gnucash.org/wiki/Online_Quotes#Finance::Quote_Sources> and 
> the following section.

For a third time. You are missing the use of a "multiple" source.
